### ISSUE TITLE
关于第二阶段主体复原部分模型训练的val数据的一些疑问?

### ISSUE THREAD

**Issue by Mmmmwei** [taco-group/MWFormer#3]
class TrainingDataset(data.Dataset):
    # type 0: raindrop;  type 1: rain;  type 2: snow;  type 3: full
    def __init__(self, type, crop_size, train_data_dir, train_filename):
        super().__init__()
        train_list = train_data_dir + train_filename    #txt filename
        with open(train_list) as f:
            contents = f.readlines()
            if type==0:
                input_names = [i[13:].strip() for i in contents if i.strip().split("/")[-1].endswith("rain.png")]  
            elif type==1:
                input_names = [i[13:].strip() for i in contents if i.split("/")[-1].startswith("im")]
            elif type==2:
                input_names = [i[13:].strip() for i in contents if i.strip().split("/")[-1].endswith("jpg")]
            else:
                input_names = [i[13:].strip() for i in contents]
            gt_names = [i.strip().replace('input','gt') for i in input_names]

以上是TrainingDataset的部分函数，该函数似乎是根据type来筛选一个数据集文件如allweather.txt中的特定天气数据或全部数据，但是下面main_train.py代码中也是从allweather.txt中抽样5%的数据作为验证集，我的问题是这样抽取的方法是否会导致训练集中的数据出现在验证集中？
   
    raindrop_dataset = TrainingDataset(0, crop_size, train_data_dir,labeled_name)
    raindrop_loader = data.DataLoader(raindrop_dataset, batch_size=10, shuffle=True, num_workers=args.num_workers,
                                        pin_memory=True, drop_last=True)    #load raindrop dataset (drop last)
    raindrop_loader_iter = enumerate(raindrop_loader)

    rain_dataset = TrainingDataset(1, crop_size, train_data_dir,labeled_name)
    rain_loader = data.DataLoader(rain_dataset,batch_size=11, shuffle=True, num_workers=args.num_workers,
                                        pin_memory=True, drop_last=True)    #load rain dataset
    rain_loader_iter = enumerate(rain_loader)

    snow_dataset = TrainingDataset(2, crop_size, train_data_dir,labeled_name)
    snow_loader = data.DataLoader(snow_dataset,batch_size=11, shuffle=True, num_workers=args.num_workers,
                                        pin_memory=True, drop_last=True)    #load snow dataset
    snow_loader_iter = enumerate(snow_loader)

   full_dataset = TrainingDataset(3, crop_size, train_data_dir,labeled_name)
    train_size = int(0.95 * len(full_dataset))
    test_size = len(full_dataset) - train_size
    _, test_dataset = torch.utils.data.random_split(full_dataset, [train_size, test_size])
    val_data_loader = data.DataLoader(test_dataset, batch_size=4, shuffle=True, num_workers=args.num_workers, pin_memory=True)      # batch size?

**Comment by zrx2000** [comment]
你好，训练过程只需要保证测试集的数据没有泄漏到训练集或验证集中即可。分出来5%的数据作为验证集算PSNR可以用于判断超参设置是否合理、训练是否在正常地进行。也可以不设置验证集，直接取最后一轮训练所保存的模型权重

**Comment by fly-dk** [comment]
> class TrainingDataset(data.Dataset): # type 0: raindrop; type 1: rain; type 2: snow; type 3: full def **init**(self, type, crop_size, train_data_dir, train_filename): super().**init**() train_list = train_data_dir + train_filename #txt filename with open(train_list) as f: contents = f.readlines() if type==0: input_names = [i[13:].strip() for i in contents if i.strip().split("/")[-1].endswith("rain.png")] elif type==1: input_names = [i[13:].strip() for i in contents if i.split("/")[-1].startswith("im")] elif type==2: input_names = [i[13:].strip() for i in contents if i.strip().split("/")[-1].endswith("jpg")] else: input_names = [i[13:].strip() for i in contents] gt_names = [i.strip().replace('input','gt') for i in input_names]
> 
> 以上是TrainingDataset的部分函数，该函数似乎是根据type来筛选一个数据集文件如allweather.txt中的特定天气数据或全部数据，但是下面main_train.py代码中也是从allweather.txt中抽样5%的数据作为验证集，我的问题是这样抽取的方法是否会导致训练集中的数据出现在验证集中？
> 
> ```
> raindrop_dataset = TrainingDataset(0, crop_size, train_data_dir,labeled_name)
> raindrop_loader = data.DataLoader(raindrop_dataset, batch_size=10, shuffle=True, num_workers=args.num_workers,
>                                     pin_memory=True, drop_last=True)    #load raindrop dataset (drop last)
> raindrop_loader_iter = enumerate(raindrop_loader)
> 
> rain_dataset = TrainingDataset(1, crop_size, train_data_dir,labeled_name)
> rain_loader = data.DataLoader(rain_dataset,batch_size=11, shuffle=True, num_workers=args.num_workers,
>                                     pin_memory=True, drop_last=True)    #load rain dataset
> rain_loader_iter = enumerate(rain_loader)
> 
> snow_dataset = TrainingDataset(2, crop_size, train_data_dir,labeled_name)
> snow_loader = data.DataLoader(snow_dataset,batch_size=11, shuffle=True, num_workers=args.num_workers,
>                                     pin_memory=True, drop_last=True)    #load snow dataset
> snow_loader_iter = enumerate(snow_loader)
> ```
> 
> full_dataset = TrainingDataset(3, crop_size, train_data_dir,labeled_name) train_size = int(0.95 * len(full_dataset)) test_size = len(full_dataset) - train_size _, test_dataset = torch.utils.data.random_split(full_dataset, [train_size, test_size]) val_data_loader = data.DataLoader(test_dataset, batch_size=4, shuffle=True, num_workers=args.num_workers, pin_memory=True) # batch size?

@Mmmmwei 你好，请问你的环境是怎么搭建的？可以让我参考一下吗？谢谢！

**Comment by Sunrise-57** [comment]
你好，我配置环境的时候也出问题了，后来用的Bi-directional Adapter for Multi-modal Tracking的环境，能直接用

> > class TrainingDataset(data.Dataset): # type 0: raindrop; type 1: rain; type 2: snow; type 3: full def **init**(self, type, crop_size, train_data_dir, train_filename): super().**init**() train_list = train_data_dir + train_filename #txt filename with open(train_list) as f: contents = f.readlines() if type==0: input_names = [i[13:].strip() for i in contents if i.strip().split("/")[-1].endswith("rain.png")] elif type==1: input_names = [i[13:].strip() for i in contents if i.split("/")[-1].startswith("im")] elif type==2: input_names = [i[13:].strip() for i in contents if i.strip().split("/")[-1].endswith("jpg")] else: input_names = [i[13:].strip() for i in contents] gt_names = [i.strip().replace('input','gt') for i in input_names]
> > 以上是TrainingDataset的部分函数，该函数似乎是根据type来筛选一个数据集文件如allweather.txt中的特定天气数据或全部数据，但是下面main_train.py代码中也是从allweather.txt中抽样5%的数据作为验证集，我的问题是这样抽取的方法是否会导致训练集中的数据出现在验证集中？
> > ```
> > raindrop_dataset = TrainingDataset(0, crop_size, train_data_dir,labeled_name)
> > raindrop_loader = data.DataLoader(raindrop_dataset, batch_size=10, shuffle=True, num_workers=args.num_workers,
> >                                     pin_memory=True, drop_last=True)    #load raindrop dataset (drop last)
> > raindrop_loader_iter = enumerate(raindrop_loader)
> > 
> > rain_dataset = TrainingDataset(1, crop_size, train_data_dir,labeled_name)
> > rain_loader = data.DataLoader(rain_dataset,batch_size=11, shuffle=True, num_workers=args.num_workers,
> >                                     pin_memory=True, drop_last=True)    #load rain dataset
> > rain_loader_iter = enumerate(rain_loader)
> > 
> > snow_dataset = TrainingDataset(2, crop_size, train_data_dir,labeled_name)
> > snow_loader = data.DataLoader(snow_dataset,batch_size=11, shuffle=True, num_workers=args.num_workers,
> >                                     pin_memory=True, drop_last=True)    #load snow dataset
> > snow_loader_iter = enumerate(snow_loader)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > full_dataset = TrainingDataset(3, crop_size, train_data_dir,labeled_name) train_size = int(0.95 * len(full_dataset)) test_size = len(full_dataset) - train_size _, test_dataset = torch.utils.data.random_split(full_dataset, [train_size, test_size]) val_data_loader = data.DataLoader(test_dataset, batch_size=4, shuffle=True, num_workers=args.num_workers, pin_memory=True) # batch size?
> 
> [@Mmmmwei](https://github.com/Mmmmwei) 你好，请问你的环境是怎么搭建的？可以让我参考一下吗？谢谢！
